### ISSUE TITLE
Feature proposal: All way to authenticate calls to custom controllers (sync/finalize/customize hooks)

### ISSUE THREAD

**Issue by grzesuav** [metacontroller/metacontroller#255]
# Request description

Currently metacontroller uses `http` communication without any authorization to execute hookhs.

# Proposal

Add ways to authenticate, like basic auth, ca certs, bearer token. 

Hook API will change, so need to carefully introduce this change to not make any breaking changes.

**Comment by MartinKanters** [comment]
Thanks for creating this. I agree that all of these means of authentication would be valuable, but these can be fixed in multiple PRs/iterations, right? Otherwise this has the risk of blowing up into a lot of work.

p.s. there's a typo in the title. All -> Add.

**Comment by grzesuav** [comment]
> Thanks for creating this. I agree that all of these means of authentication would be valuable, but these can be fixed in multiple PRs/iterations, right? Otherwise this has the risk of blowing up into a lot of work.

it depends, I want first check how it can be implemented, potentially there will be an easy way to do it in one go, and if not - at least API need to be defined/proposed in first PR, even if it won't be implemented to reduce number of changes in CRD's definition. Of course, implementation of methods can be separated if needed.

> 
> p.s. there's a typo in the title. All -> Add.

thanks :)

**Comment by MartinKanters** [comment]
> it depends, I want first check how it can be implemented, potentially there will be an easy way to do it in one go, and if not - at least API need to be defined/proposed in first PR, even if it won't be implemented to reduce number of changes in CRD's definition. Of course, implementation of methods can be separated if needed.

Makes sense! Thanks, looking forward to it.

**Comment by grzesuav** [comment]
I am thinking of having sth similar as in https://github.com/prometheus/common/blob/main/config/http_config.go#L159 - @MartinKanters do you think it cover most common use cases ?

**Comment by grzesuav** [comment]
Or similar https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/apis/monitoring/v1/types.go#L665-L675

**Comment by MartinKanters** [comment]
Took a quick look at it and it looks fine to me!

**Comment by dkhacharia** [comment]
I am also waiting for this feature to be able to adopt metacontroller. Is it in the making?

**Comment by grzesuav** [comment]
not yet, focusing on other changes currently. Also, currently touching the logic around the hooks.

**Comment by NiklasRosenstein** [comment]
Is the project still open to a contribution on this?

Maybe we can add a secret ref to the `webhook:` field?

**Comment by grzesuav** [comment]
@NiklasRosenstein yes, open. If you have an idea an you either write it uphere or start a thread in metacontroller kubernetes channel on slack ?

**Comment by flavio** [comment]
I'm doing an evaluation of metacontroller and I run into this missing feature that I would need. I've opened https://github.com/metacontroller/metacontroller/pull/1204 to address that. I hope you won't mind 😅
